fix(carto): Clean up and add unit tests for requestWithParameters cache

donmccurdy opened the pull request:
Related:

- #8648

Adds unit tests for `requestWithParameters`.

`requestWithParameters()` does not include the given `accessToken` in the request, but appends it to the response body. This behaves confusingly with the cache — a second caller may get the first caller's access token — and only one caller to `requestWithParameters()` currently provides that parameter, so I've removed the access token parameter and applied it in the calling function instead.

***

OUTDATED: ~~While requestWithParameters() caches unique requests, it should apply accessToken and errorContext separately for each invocation, regardless of whether separate requests were made.~~

donmccurdy commented:
* **#8707** <a href="https://app.graphite.dev/github/pr/visgl/deck.gl/8707?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @donmccurdy and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

felixpalmer commented:
Why is this required? It seems surprising to me that if the client makes a request with a valid token, and then again with an invalid one that the latter will succeed.

donmccurdy commented:
Note that `accessToken` is added to the response — not the request — unless the caller has included an access token in the request headers explicitly as well. If `accessToken` is part of the request headers then the two requests are distinct, and do not share a cache key.

This is a bit confusing. Perhaps it would be better to remove the `accessToken` argument from the function entirely. It appears to be used only in one place ...

https://github.com/visgl/deck.gl/blob/d9a78daf4bf3a41a2b8b01fe8f768ff8e1cee741/modules/carto/src/sources/base-source.ts#L58-L65

... so we could append `accessToken` to the response there, instead? I'm tempted to do that.

***

Probably the caller expects receive the error context they originally provided, so I suspect that part of the change should remain.

***

That said – I don't know that requestWithParameter *needs* to behave one way or another, if we prefer the previous behavior. But having the unit tests in place should make it easier to see what it's doing, which I did not when changing the cache behavior in #8648.

felixpalmer reviewed (changes requested):
Thanks for the explanation. 

Now I remember why we do this. We need to have `requestWithParameters` return the exact object in the cache so that deck doesn't think it has a new `data` object. This is important for the React use-case where in a `render` function the user can define the source and expect things to work.

So I don't think we can make a change where the object that is retrieved from the cache is modified in any way

felixpalmer commented:
We need to be able to support this pattern:

```js
function MyComponent(tableName) {
  const data = vectorTableSource({...globalOptions, tableName});

  return new VectorTileLayer({
    data
    ...
  });
}
```

If `data` is not equal between frames, deck will keep recreating the layer

donmccurdy commented:
Ok, that makes sense to me — I'll revert the `accessToken` changes then, and keep the error handling and the added unit tests.

donmccurdy commented:
@felixpalmer I've reverted the changes to `requestWithParameters()` keeping just the unit tests. Because the `accessToken` parameter was only used in one place, leaks to subsequent requests, and brings in the concerns you mentioned about layer rendering performance, I've removed that argument and assigned it from the calling function instead.
